### PR TITLE
Floating point comparison with ULP

### PR DESCRIPTION
**Pull request by yxlao**
- Added `close_f` for comparing floating points
- Only converted selected unit test to use `all_close_f` for NNP

### CODE REVIEW

File: test/autodiff.in.cpp [hunk 1]
```diff
@@ -1364,7 +1364,7 @@ NGRAPH_TEST(${BACKEND_NAME}, backwards_tan)
     auto pi = 3.14159f;
 
     // Stay away from the asymptotes at 6 and 12 o'clock.
-    auto slop = 0.1f;
+    auto slop = 0.2f;
```
Comment (yxlao) [inline test/autodiff.in.cpp]: To stay further away from 6 and 12 o'clock

Comment (aprocter) [reply test/autodiff.in.cpp]: No objection from me... `0.1f` was kind of arbitrary in the first place.

File: test/util/all_close.hpp [hunk 2]
```diff
@@ -39,15 +39,17 @@ namespace ngraph
...
                        T atol = static_cast<T>(1e-8))
         {
+            bool rc = true;
             assert(a.size() == b.size());
             for (size_t i = 0; i < a.size(); ++i)
             {
                 if (std::abs(a[i] - b[i]) > atol + rtol * std::abs(b[i]))
                 {
-                    return false;
+                    NGRAPH_INFO << a[i] << " !≈ " << b[i];
```
Comment (aprocter) [inline test/util/all_close.hpp]: Minor nit: we should probably stay within 7-bit ASCII, which precludes using `≈`

File: test/util/all_close_f.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    if (a.size() != b.size())
+    {
+        throw ngraph_error("a.size() != b.size() for all_close comparison.");
+    }
+    for (size_t i = 0; i < a.size(); ++i)
+    {
+        bool is_close_f = close_f(a[i], b[i], mantissa_bits, tolerance_bits);
+        if (!is_close_f)
+        {
+            NGRAPH_INFO << a[i] << " !≈ " << b[i];
```
Comment (aprocter) [inline test/util/all_close_f.cpp]: Same minor nit as above on use of `≈`

File: test/util/test_tools.cpp [hunk 4]
```diff
@@ -23,6 +23,74 @@
...
+        float_vec = vector<float>(vec.begin(), vec.end());
+    }
+    else if (element_type == element::u64)
+    {
+        vector<uint64_t> vec = read_vector<uint64_t>(tv);
+        float_vec = vector<float>(vec.begin(), vec.end());
+    }
+    else
+    {
+        throw ngraph_error("Unsupported ngraph element type.");
```
Comment (aprocter) [inline test/util/test_tools.cpp]: n`G`raph :)

**Comment by yxlao** [comment]
good point @ransford2011 , added `nan` and `inf` handling

**Pull request merged**